src/twisted/conch/ssh/transport.py

```diff
@@ -730,34 +730,56 @@ def dataReceived(self, data):
         self.buf = self.buf + data
         if not self.gotVersion:
 
-            if len(self.buf) > 4096:
+            if len(self.buf) >= 255 and b"\n" not in self.buf:
+                # Looks like we have already received enough data,
+                # yet it doesn't look like the version string line
+                # is in the received data.
+                # RFC 4253 section 4.2
+                # The maximum length of the string is 255 characters,
+                # including the Carriage Return and Line Feed.
                 self.sendDisconnect(
                     DISCONNECT_CONNECTION_LOST,
-                    b"Peer version string longer than 4KB. "
-                    b"Preventing a denial of service attack.",
+                    b"Invalid peer version. Long value. Preventing a denial of service attack.",
                 )
                 return
 
-            if self.buf.find(b"\n", self.buf.find(b"SSH-")) == -1:
+            if b"\n" not in self.buf:
+                # We don't have yet the full version string.
+                return
+
+            if not self.buf.startswith(b"SSH-"):
+                # The identification string MUST be
+                # SSH-protoversion-softwareversion SP comments CR LF
+                self.sendDisconnect(
+                    DISCONNECT_CONNECTION_LOST, b"Invalid peer version format."
+                )
```

Comment:
This seems to violate a "MUST" in RFC 4253 section 4.2, in the case where the peer is a server:

```
   The server MAY send other lines of data before sending the version
   string.  Each line SHOULD be terminated by a Carriage Return and Line
   Feed.  Such lines MUST NOT begin with "SSH-", and SHOULD be encoded
   in ISO-10646 UTF-8 [RFC3629] (language is not specified).  Clients
   MUST be able to process such lines.  Such lines MAY be silently
   ignored, or MAY be displayed to the client user.  If they are
   displayed, control character filtering, as discussed in [SSH-ARCH],
   SHOULD be used.  The primary use of this feature is to allow TCP-
   wrappers to display an error message before disconnecting.
```

It would be OK to accept lines before the identification string only in the client case, though; and it may be permissible to limit the number of such lines even in the client case.  For example, OpenSSH does this, where `SSH_MAX_PRE_BANNER_LINES` is defined to 1024:

```c
                /* Accept lines before banner only on client */
                if (ssh->kex->server || ++n > SSH_MAX_PRE_BANNER_LINES) {
  bad:
                        if ((r = sshbuf_put(ssh_packet_get_output(ssh),
                            mismatch, strlen(mismatch))) != 0)
                                return r;
                        return SSH_ERR_NO_PROTOCOL_VERSION;
                }
```

Clients don't have the same DoS concerns that servers have.

Reply:
Many thanks. I missed that part of the RFC.

I saw that there was an explicit test for this behaviour but I didn't knew why.

I will revert and document this. Thanks!

Reply:
> It would be OK to accept lines before the identification string only in the client case, though; and it may be permissible to limit the number of such lines even in the client case. For example, OpenSSH does this, where SSH_MAX_PRE_BANNER_LINES is defined to 1024:

previously  SSHTransportBase [buffered the data then dropped the lines](https://github.com/twisted/twisted/pull/1697/files#diff-92b57e49a6a90fe8379778df32a08366febe251584794e1aa01ffd878a674690L748-L750), when restoring this behaviour it would be better if any data that doesn't start b"SSH-" is dropped immediately